File: atoms-ui/harness/transport/contracts.ts
export type AtomId = string;
export type CanvasId = string;
export type ActorId = string;
export type Revision = number;

// --- Event Type Constants (REALTIME_SPEC_V1) ---

export const EventType = {
    // Downstream (SSE)
    CANVAS_READY: 'CANVAS_READY',
    TOKEN_PATCH: 'token_patch',
    STATE_PATCH: 'state_patch',
    CANVAS_COMMIT: 'canvas_commit',
    SNAPSHOT_CREATED: 'snapshot_created',
    CHAT_TOKEN: 'chat_token',
    LOG_LINE: 'log_line',

    // Presence/Ephemeral
    PRESENCE_STATE: 'presence_state',
    GESTURE: 'gesture',

    // Legacy/Chat
    USER_MESSAGE: 'user_message',
    AGENT_MESSAGE: 'agent_message',
    RESUME_CURSOR: 'resume_cursor',
} as const;

export type EventTypeValue = typeof EventType[keyof typeof EventType];

// --- Routing Keys ---

export type ActorType = 'human' | 'agent' | 'system';

export interface RoutingKeys {
    tenant_id: string;
    mode?: string;
    env?: 'dev' | 'staging' | 'prod' | 'stage';
    workspace_id?: string;
    project_id: string;
    app_id?: string;
    surface_id?: string;
    canvas_id?: string;
    projection_id?: string;
    panel_id?: string;
    thread_id?: string;
    actor_id: string;
    actor_type: ActorType;
    session_id?: string;
    device_id?: string;
}

export interface EventIds {
    request_id?: string;
    correlation_id?: string;
    run_id?: string;
    step_id?: string;
}

// --- Atoms & Tokens ---

export type PrimitiveValue = string | number | boolean | null;

export interface TokenValue {
    type: 'static' | 'bound';
    value: PrimitiveValue;
    fallback?: PrimitiveValue;
    meta?: Record<string, unknown>; // tracking, schema, etc.
}

export interface Atom {
    id: AtomId;
    type: string; // e.g., 'text', 'box', 'image'
    parent_id: AtomId | null; // null for root atoms
    properties: Record<string, TokenValue | PrimitiveValue>;
    children?: AtomId[]; // Ordered list of child IDs
}

// --- Operations (Mutations) ---

export type OpType = 'add' | 'remove' | 'update' | 'move';

export interface BaseOp {
    kind: OpType;
    actor_id: ActorId;
}

export interface AddOp extends BaseOp {
    kind: 'add';
    atom: Atom;
    parent_id: AtomId | null; // Where to add
    index: number; // Insert position
}

export interface RemoveOp extends BaseOp {
    kind: 'remove';
    atom_id: AtomId;
}

export interface UpdateOp extends BaseOp {
    kind: 'update';
    atom_id: AtomId;
    properties: Record<string, TokenValue | PrimitiveValue>; // Partial update
}

export interface MoveOp extends BaseOp {
    kind: 'move';
    atom_id: AtomId;
    new_parent_id: AtomId | null;
    new_index: number;
}

export type CanvasOp = AddOp | RemoveOp | UpdateOp | MoveOp;

// --- Commands (Transport) ---

export interface Command {
    base_rev: Revision;
    ops: CanvasOp[];
    actor_id: ActorId;
    correlation_id: string; // For idempotency & optimistic updates

    // Generic Extensions for Tool Control
    type?: string;
    command?: string;
    payload?: any;
    context?: any;
}

export interface CommandResponse {
    success: boolean;
    head_rev: Revision;
    applied_ops?: CanvasOp[];
    error?: {
        code: string;
        message: string;
        expected_rev?: Revision;
        recovery_ops?: CanvasOp[]; // For conflict resolution
    };
}

// --- Realtime / Stream Events ---

export type GestureKind = 'caret' | 'drag' | 'cursor';

export interface GesturePayload {
    kind: GestureKind;
    atom_id?: AtomId; // For selection/caret
    position?: { x: number; y: number }; // For cursors/drags
    meta?: unknown;
    // Generic
    type?: string;
    x?: number;
    y?: number;
    focus_id?: string;
}

export interface GestureEvent {
    type: 'gesture';
    data: {
        kind?: GestureKind; // Optional flexibility
        payload?: GesturePayload;
        actor_id?: ActorId;
        // Flattened structure support
        type?: string;
        x?: number;
        y?: number;
        focus_id?: string;
    };
}

export interface OpCommittedEvent {
    type: 'op_committed';
    data: {
        rev: Revision;
        ops: CanvasOp[];
        actor_id: ActorId;
    };
}

export interface SystemEvent {
    type: 'system';
    data: {
        code: string;
        message: string;
    };
}

export interface SafetyDecisionEvent {
    type: 'safety_decision';
    data: {
        action: string;
        result: 'PASS' | 'BLOCK';
        reason: string;
        gate: string;
        [key: string]: unknown;
    };
    routing?: {
        canvas_id?: string;
        thread_id?: string;
        [key: string]: unknown;
    };
    ids?: {
        run_id?: string;
        [key: string]: unknown;
    };
    meta?: {
        [key: string]: unknown;
    };
}

export interface MediaSidecar {
    uri?: string;
    object_id?: string;
    artifact_id?: string;
    mime_type?: string;
    size_bytes?: number;
    checksum?: string;
    metadata?: Record<string, unknown>;
}

/**
 * Validate MediaSidecar (client-side check).
 * Rejects data: URIs and base64 content per REALTIME_SPEC_V1 §1.5.
 */
export function validateMediaSidecar(sidecar: MediaSidecar): void {
    if (!sidecar.uri && !sidecar.object_id && !sidecar.artifact_id) {
        throw new Error('MediaSidecar requires uri, object_id, or artifact_id');
    }

    if (sidecar.uri) {
        if (sidecar.uri.startsWith('data:')) {
            throw new Error('data: URIs are forbidden in media sidecars; use durable references (s3://, artifact_id, etc.)');
        }
        if (sidecar.uri.includes('base64')) {
            throw new Error('base64-encoded content is forbidden in media sidecars');
        }
    }
}

export interface MediaPayload {
    sidecars: MediaSidecar[];
    caption?: string;
}

export interface StreamEnvelopeBase {
    type: string;
    atom_metadata?: Record<string, unknown>;
    media_payload?: MediaPayload;
    [key: string]: unknown;
}

export interface ChatMessage {
    id: string;
    role: 'user' | 'assistant' | 'system';
    content: string;
    timestamp: string;
}

export type StreamEvent =
    | (GestureEvent & StreamEnvelopeBase)
    | (OpCommittedEvent & StreamEnvelopeBase)
    | (SystemEvent & StreamEnvelopeBase)
    | (SafetyDecisionEvent & StreamEnvelopeBase);

File: atoms-core/src/atoms_core/realtime/contracts.py
"""Canonical Realtime Event Contracts (REALTIME_SPEC_V1)."""
from __future__ import annotations

import uuid
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Dict, List, Literal, Optional, Union

from pydantic import BaseModel, Field, root_validator, validator

# --- Routing Keys ---

class ActorType(str, Enum):
    HUMAN = "human"
    AGENT = "agent"
    SYSTEM = "system"


class RoutingKeys(BaseModel):
    """
    Mandatory routing keys for all StreamEvent envelopes.
    """
    tenant_id: str = Field(..., pattern=r"^t_[a-z0-9_-]+$")
    mode: Optional[str] = None
    env: Literal["dev", "staging", "prod", "stage"] = "dev"
    # Hierarchy
    workspace_id: Optional[str] = None
    project_id: str = Field(..., min_length=1)
    app_id: Optional[str] = None
    # Surface
    surface_id: Optional[str] = None
    canvas_id: Optional[str] = None
    projection_id: Optional[str] = None
    panel_id: Optional[str] = None
    thread_id: Optional[str] = None
    # Actor
    actor_id: str
    actor_type: ActorType
    session_id: Optional[str] = None
    device_id: Optional[str] = None


# --- IDs ---

class EventIds(BaseModel):
    request_id: Optional[str] = None
    correlation_id: Optional[str] = None
    run_id: Optional[str] = None
    step_id: Optional[str] = None


# --- Priority & Meta ---

class EventPriority(str, Enum):
    TRUTH = "truth"  # Authoritative commits
    GESTURE = "gesture"  # Ephemeral (mouse, typing)
    INFO = "info"  # Notifications/Chat
    CRITICAL = "critical"  # Severe errors/interrupts


class PersistPolicy(str, Enum):
    ALWAYS = "always"
    SAMPLED = "sampled"
    NEVER = "never"


class EventSeverity(str, Enum):
    INFO = "info"
    # Ported minimal enum to avoid dependency
    DEBUG = "debug"
    WARN = "warn"
    ERROR = "error"


class StorageClass(str, Enum):
    STREAM = "stream"
    LOG = "log"


class EventMeta(BaseModel):
    schema_ver: Optional[str] = None
    priority: EventPriority = EventPriority.INFO
    persist: PersistPolicy = PersistPolicy.ALWAYS
    last_event_id: Optional[str] = None
    schema_version: str = "1.0.0"
    severity: EventSeverity = Field(default=EventSeverity.INFO)
    storage_class: StorageClass = Field(default=StorageClass.STREAM)


# --- Payload Helpers ---

class MediaSidecar(BaseModel):
    """
    Sidecar reference for media payloads (URLs, object IDs, artifacts).
    Raw bytes or data URIs are explicitly rejected.
    """
    uri: Optional[str] = None
    object_id: Optional[str] = None
    artifact_id: Optional[str] = None
    mime_type: Optional[str] = None
    size_bytes: Optional[int] = None
    checksum: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)

    @root_validator(skip_on_failure=True)
    def _require_reference(cls, values: dict[str, Any]) -> dict[str, Any]:
        uri = values.get("uri")
        object_id = values.get("object_id")
        artifact_id = values.get("artifact_id")
        if not (uri or object_id or artifact_id):
            raise ValueError("media sidecar requires uri, object_id, or artifact_id")
        
        # Reject base64 and data URIs (REALTIME_SPEC_V1 §1.5)
        if uri:
            if uri.startswith("data:"):
                raise ValueError("data: URIs are forbidden in media sidecars; use durable references (s3://, artifact_id, etc.)")
            if uri.startswith("base64,") or "base64" in uri.lower():
                raise ValueError("base64-encoded content is forbidden in media sidecars")
        
        return values


class MediaPayload(BaseModel):
    sidecars: List[MediaSidecar] = Field(default_factory=list)
    caption: Optional[str] = None
    atom_metadata: Dict[str, Any] = Field(default_factory=dict)


class AtomMetadataPayload(BaseModel):
    atom_id: str
    atom_revision: Optional[Union[str, int]] = None
    atom_metadata: Dict[str, Any] = Field(default_factory=dict)
    media_payload: Optional[MediaPayload] = None
    surface_id: Optional[str] = None
    canvas_id: Optional[str] = None
    projection_id: Optional[str] = None
    panel_id: Optional[str] = None


# --- Canonical StreamEvent ---

class StreamEvent(BaseModel):
    """
    The Single Source of Truth event envelope.
    """
    v: int = 1
    type: str  # e.g. "token_chunk", "canvas_commit", "presence_state"
    ts: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    seq: Optional[int] = None  # Monotonic sequence per stream key
    event_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    trace_id: Optional[str] = None
    span_id: Optional[str] = None

    ids: EventIds = Field(default_factory=EventIds)
    routing: RoutingKeys
    data: Dict[str, Any] = Field(default_factory=dict)
    atom_metadata: Dict[str, Any] = Field(default_factory=dict)
    media_payload: Optional[MediaPayload] = None
    meta: EventMeta = Field(default_factory=EventMeta)

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


# --- Chat Types (Merged) ---

class Contact(BaseModel):
    id: str
    display_name: Optional[str] = None
    handle: Optional[str] = None


class Message(BaseModel):
    id: str
    thread_id: str
    sender: Contact
    text: str
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    role: str = "user"  # user | agent | system
    scope: Optional[Dict[str, Any]] = None


# --- Event Type Constants (REALTIME_SPEC_V1 §1.7) ---

class EventType:
    """Canonical event type constants."""
    # Downstream (SSE)
    CANVAS_READY = "CANVAS_READY"
    TOKEN_PATCH = "token_patch"
    STATE_PATCH = "state_patch"
    CANVAS_COMMIT = "canvas_commit"
    SNAPSHOT_CREATED = "snapshot_created"
    CHAT_TOKEN = "chat_token"
    LOG_LINE = "log_line"
    
    # Presence/Ephemeral
    PRESENCE_STATE = "presence_state"
    GESTURE = "gesture"
    
    # Legacy/Chat
    USER_MESSAGE = "user_message"
    AGENT_MESSAGE = "agent_message"
    RESUME_CURSOR = "resume_cursor"


# --- Command Envelope (Upstream) ---

class CommandEnvelope(BaseModel):
    """
    Upstream command envelope (HTTP/WS).
    Commands are idempotent, typed, and scoped by routing.
    """
    command_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    command_type: str  # e.g. "canvas.atom.create", "chat.send"
    routing: RoutingKeys
    payload: Dict[str, Any] = Field(default_factory=dict)
    idempotency_key: Optional[str] = None
    ts: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


# --- Event Envelope (Downstream) ---

class EventEnvelope(BaseModel):
    """
    Generic downstream event wrapper.
    Alias for StreamEvent with stricter typing.
    """
    event: StreamEvent
    

class DatasetEvent(BaseModel):
    """
    Event carrying dataset/artifact references.
    Used for snapshot_created, artifact uploads, etc.
    """
    event_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    event_type: str = EventType.SNAPSHOT_CREATED
    routing: RoutingKeys
    dataset_id: Optional[str] = None
    artifact_refs: List[MediaSidecar] = Field(default_factory=list)
    metadata: Dict[str, Any] = Field(default_factory=dict)
    ts: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
